mvcc/kvstore.go

```diff
@@ -271,9 +271,13 @@ func (s *store) Compact(rev int64) (<-chan struct{}, error) {
 			return
 		}
 		close(ch)
-		s.keepMu.Lock()
-		s.keep = nil
-		s.keepMu.Unlock()
+		s.revMu.RLock()
+		if rev == s.compactMainRev {
+			s.keepMu.Lock()
+			s.keep = nil
```

Comment:
That seems too delicate. It might be easier to extend the `index` interface to have a non-destructive Compact which computes the maximum revisions less than a given rev (e.g., the compact rev) and just run that every time HashKV is called.